float32 exf_x
float32 exf_y
float32 exf_z
